homeassistant/components/tplink/__init__.py

```diff
@@ -126,3 +201,31 @@ def legacy_device_id(device: SmartDevice) -> str:
     if "_" not in device_id:
         return device_id
     return device_id.split("_")[1]
+
+
+def _get_store(hass: HomeAssistant) -> Store[dict[str, dict[str, str]]]:
+    if DOMAIN in hass.data and DATA_STORE in hass.data[DOMAIN]:
+        return cast(Store[dict[str, dict[str, str]]], hass.data[DOMAIN][DATA_STORE])
+    return Store[dict[str, dict[str, str]]](hass, STORAGE_VERSION, STORAGE_KEY)
+
+
+async def get_stored_credentials(hass: HomeAssistant) -> Optional[Credentials]:
+    """Retrieve the credentials from the Store."""
+    storage_data = await _get_store(hass).async_load()
```

Comment:
This means every device is going to re-read the file from disk

Reply:
Fixed to now cache credentials in hass.data and get from there.